numba/npyufunc/dufunc.py

```diff
@@ -0,0 +1,48 @@
+from __future__ import absolute_import, print_function, division
+import numpy
+
+from .. import jit, typeof, numpy_support, utils
+from . import _internal, ufuncbuilder
+
+class DUFunc(_internal._DUFunc):
+    # NOTE: __base_kwargs must be kept in synch with the kwlist in
+    # _internal.c:dufunc_init()
+    __base_kwargs = set(('identity', 'keepalive', 'nin', 'nout'))
+
+    def __init__(self, py_func, **kws):
+        dispatcher = jit(target='npyufunc')(py_func)
+        self.targetoptions = {}
+        # Loop over a copy of the keys instead of the keys themselves,
+        # since we're changing the dictionary while looping.
+        kws_keys = tuple(kws.keys())
+        for key in kws_keys:
+            if key not in self.__base_kwargs:
+                self.targetoptions[key] = kws.pop(key)
+        kws['identity'] = ufuncbuilder._BaseUFuncBuilder.parse_identity(
+            kws.pop('identity', None))
+        super(DUFunc, self).__init__(dispatcher, **kws)
+
+    def _compile_for_args(self, *args, **kws):
+        assert len(kws) == 0
+        argtys = []
+        # To avoid a mismatch in how Numba types values as opposed to
+        # Numpy, we need to first check for scalars.  For example, on
+        # 64-bit systems, numba.typeof(3) => int32, but
+        # numpy.array(3).dtype => int64.
+        for arg in args[:self.ufunc.nin]:
+            if numpy_support.is_arrayscalar(arg):
+                argtys.append(numpy_support.map_arrayscalar_type(arg))
+            else:
+                argty = typeof(arg)
```

Comment:
The intention here was to be as generic as possible.  That being said, I will modify the next line to only check for array types.  I can't think off the top of my head where the argument type will not a Numpy scalar nor an array, but I don't want to disallow that.